### PR TITLE
[WIP] More performant std.array.Appender

### PR DESCRIPTION
**Pull request by JackStouffer**
This is a revival of https://github.com/dlang/phobos/pull/502

This is a work in progress.

[patch] std.array.Appender has severe performance and memory leak problems
http://d.puremagic.com/issues/show_bug.cgi?id=5813

CC @andralex @DmitryOlshansky

### CODE REVIEW

**Comment by CyberShadow** [comment]
Needs rationale / explanation why this introduces a new type instead of improving `Appender`, migration/deprecation path...

**Comment by JackStouffer** [comment]
Still working on that. I just changed the name because Andrei requested it. Going through the review on the original PR now.

I think he wanted the uses of the two become orthogonal.

**Comment by CyberShadow** [comment]
Does the API *have* to change, though? I understand that the old appender keeps everything in one continuous block whereas this one has chunks, but that doesn't mean that e.g. it couldn't coalesce them on-demand when `.data` is accessed.

**Comment by DmitryOlshansky** [comment]
> it couldn't coalesce them on-demand when .data is accessed.

This has subtle performance issues all over it. Consider e.g. a stack based on old appender (a plausible design) the user called `.data` freely on each access to the top of the stack. With auto-magically coalescing this will be quite expensive.

**Comment by CyberShadow** [comment]
> This has subtle performance issues all over it. Consider e.g. a stack based on old appender (a plausible design) the user called .data freely on each access to the top of the stack. With auto-magically coalescing this will be quite expensive.

Not necessarily, as the coalescing might preallocate in the same way. We could provide a new method which coalesces without preallocation instead, though.

**Comment by DmitryOlshansky** [comment]
> Not necessarily, as the coalescing might preallocate in the same way. We could provide a new method which coalesces without preallocation instead, though.

It is still an extra check on each access to `.data` it might be 100% predictable though.

**Comment by CyberShadow** [comment]
> It is still an extra check on each access to .data it might be 100% predictable though.

I think that's completely fine if we get performance gains elsewhere. What's important is that the (amortized) cyclomatic complexity doesn't change.

**Comment by JackStouffer** [comment]
My benchmarks show this is at lest three times slower than `Appender`.

```
appender	87 ms, 566 μs, and 1 hnsec
builder		319 ms, 418 μs, and 3 hnsecs
```

Let me know if you think this benchmark is unfair in some way

```d
import std.array;
import std.traits;
import std.stdio;
import std.file;
import std.datetime;
import std.conv;
import std.range;
import std.algorithm;
import std.random;

enum testCount = 10_000;

__gshared int[] random1 = [];
__gshared int[] random2 = [];
__gshared int[] random3 = [];


auto appenderTest()()
{
    Appender!(int[]) app;
    app ~= random1;
    app ~= random2;
    app ~= random3;

    return app.data;
}


auto builderTest()()
{
    Builder!(int[]) builder;
    builder ~= random1;
    builder ~= random2;
    builder ~= random3;

    return builder[];
}


void main()
{

    foreach (_; 0 .. 10_000)
        random1 ~= uniform(0, 10000);
    foreach (_; 0 .. 6000)
        random1 ~= uniform(0, 10000);
    foreach (_; 0 .. 1000)
        random1 ~= uniform(0, 10000);

    auto result  = to!Duration(benchmark!(() => appenderTest())(testCount)[0]);
    auto result2 = to!Duration(benchmark!(() => builderTest())(testCount)[0]);

    writeln("appender", "\t", result);
    writeln("builder", "\t\t", result2);
}
```

**Comment by JackStouffer** [comment]
Also, making the builder allocate enough space for the elements before appending them made the builder version even slower.

**Comment by CyberShadow** [comment]
Thanks for the benchmark @JackStouffer .

I vaguely recall that Appender has received a number of performance improvements since the original Builder code here was written.

IIRC, if we were to add a whole new type, you can get even more performance by abandoning ref-counting, which removes one level of indirection (to the "impl").

FWIW, I've tried a bunch of appender designs while working on DFeed. This code is ancient by current means but here's the repo: https://github.com/CyberShadow/DAppenderResearch

And here's the appender I use currently: https://github.com/CyberShadow/ae/blob/master/utils/appender.d

Probably one of its more distinct features is that its `put` can take several arguments, which allows it to check for necessary space only once per call, instead of once per argument.

**Comment by DmitryOlshansky** [comment]
```
 foreach (_; 0 .. 10_000)
        random1 ~= uniform(0, 10000);
    foreach (_; 0 .. 6000)
        random1 ~= uniform(0, 10000);
    foreach (_; 0 .. 1000)
        random1 ~= uniform(0, 10000);
```
I guess you wanted to append to random1, random2 and random3 instead.

**Comment by JackStouffer** [comment]
To be honest, looking at the design of Appender vs this, I don't see how this can be more performant in my benchmark.

The number of allocations that Appender has to do in order to append an item of 10,000 elements is just one, while Builder must do at least nine. Appender than can also assign all items in an array to it's new memory in one go while Builder must assign the array to its managed memory in nine different steps.

It's entirely possible that builder performs better when doing many arrays of a small number of elements. I will test this out.

**Comment by DmitryOlshansky** [comment]
@JackStouffer Another thing is since this benchmark is allocating stuff it is not good idea to run one benchmark after another - the heap gets fragmented by the first guy. Also try `GC.disable` to nullify GC pauses.

**Comment by JackStouffer** [comment]
@DmitryOlshansky Thanks, will test

**Comment by JackStouffer** [comment]
New code and results. Small array tests to follow.

This raises an interesting question however: if I need to disable collections in order to get good results, doesn't that speak to a GC trashing problem in builder? Users might be using multiple Builders in their code, we have to consider that Builder will cause more collections than Appender.

```
appender	336 ms, 371 μs, and 4 hnsecs
builder		336 ms, 280 μs, and 7 hnsecs
```

```d
import std.array;
import std.traits;
import std.stdio;
import std.file;
import std.datetime;
import std.conv;
import std.range;
import std.algorithm;
import std.random;

import core.memory : GC;

enum testCount = 10_000;

__gshared int[] random1 = [];
__gshared int[] random2 = [];
__gshared int[] random3 = [];


auto appenderTest()()
{
    Appender!(int[]) app;
    app ~= random1;
    app ~= random2;
    app ~= random3;

    return app.data;
}


auto builderTest()()
{
    Builder!(int[]) builder;
    builder ~= random1;
    builder ~= random2;
    builder ~= random3;

    return builder[];
}


void main()
{
    GC.disable();

    foreach (_; 0 .. 10_000)
        random1 ~= uniform(0, 10_000);
    foreach (_; 0 .. 6_000)
        random2 ~= uniform(0, 10_000);
    foreach (_; 0 .. 1_000)
        random3 ~= uniform(0, 10_000);

    auto result  = to!Duration(benchmark!(() => appenderTest())(testCount)[0]);
    auto result2 = to!Duration(benchmark!(() => builderTest())(testCount)[0]);

    writeln("appender", "\t", result);
    writeln("builder", "\t\t", result2);
}
```

**Comment by JackStouffer** [comment]
Also note that I'm returning `builder[]` because `builder.data` is so slow.

**Comment by DmitryOlshansky** [comment]
> This raises an interesting question however: if I need to disable collections in order to get good results, doesn't that speak to a GC trashing problem in builder? Users might be using multiple Builders in their code, we have to consider that Builder will cause more collections than Appender.

It's not to get good results, it's to avoid counting GC pause into the time of any of the algorithms.

**Comment by JackStouffer** [comment]
> It's not to get good results, it's to avoid counting GC pause into the time of any of the algorithms.

But that's my point. Doesn't it follow that builder is disproportionately affected by GC pauses than Appender, seeing as how after I disabled them, performance increased by 400%?

**Comment by DmitryOlshansky** [comment]
> Doesn't it follow that builder is disproportionately affected by GC pauses than Appender, seeing as how after I disabled them, performance increased by 400%?

You could also choose to run Builder first and then Append with GC enabled and see who wins.

**Comment by JackStouffer** [comment]
Changed code

```d
void main()
{
    GC.disable();

    foreach (_; 0 .. 10_000)
        random1 ~= uniform(0, 10_000);
    foreach (_; 0 .. 6_000)
        random2 ~= uniform(0, 10_000);
    foreach (_; 0 .. 1_000)
        random3 ~= uniform(0, 10_000);

    auto result2 = to!Duration(benchmark!(() => builderTest())(testCount)[0]);
    writeln("builder", "\t\t", result2);

    GC.disable();

    auto result  = to!Duration(benchmark!(() => appenderTest())(testCount)[0]);
    writeln("appender", "\t", result);
}
```

No change in speed for Appender

```
builder		330 ms, 399 μs, and 7 hnsecs
appender	336 ms, 335 μs, and 6 hnsecs
```

**Comment by DmitryOlshansky** [comment]
> Changed code

I mean don't disable GC at all and see if the ordering of Builder/Appender has effect on the results.

**Comment by JackStouffer** [comment]
Yeah, it looks like Appender is much more GC friendly than Builder

```
builder		327 ms, 977 μs, and 2 hnsecs
appender	74 ms, 806 μs, and 3 hnsecs
```

```d
void main()
{
    GC.disable();

    foreach (_; 0 .. 10_000)
        random1 ~= uniform(0, 10_000);
    foreach (_; 0 .. 6_000)
        random2 ~= uniform(0, 10_000);
    foreach (_; 0 .. 1_000)
        random3 ~= uniform(0, 10_000);

    auto result2 = to!Duration(benchmark!(() => builderTest())(testCount)[0]);
    writeln("builder", "\t\t", result2);

    GC.enable();

    auto result  = to!Duration(benchmark!(() => appenderTest())(testCount)[0]);
    writeln("appender", "\t", result);
}
```

**Comment by JackStouffer** [comment]
@DmitryOlshansky Sorry, I deleted that comment, as I made a mistake in the code

**Comment by JackStouffer** [comment]
> I mean don't disable GC at all and see if the ordering of Builder/Appender has effect on the results.

Yup, just changing the ordering and removing the disable/enable calls to the GC had the same results

```
builder		336 ms, 514 μs, and 3 hnsecs
appender	76 ms, 601 μs, and 2 hnsecs
```

File: std/array.d [hunk 1]
```diff
@@ -3559,3 +3559,1006 @@ RefAppender!(E[]) appender(P : E[]*, E)(P arrayPtr)
...
+                    return s;
+                }
+                auto seg = cast(Segment*) freelist;
+                if (seg)
+                { // There was a segment to re-use
+                    freelist = seg.next;
+                }
+                else
+                { // Allocate a new segment
+                    seg = cast(Segment*) malloc(PageSize);
```
Comment (DmitryOlshansky) [inline std/array.d]: Soo this is C's library malloc...

File: std/array.d [hunk 2]
```diff
@@ -3559,3 +3559,1006 @@ RefAppender!(E[]) appender(P : E[]*, E)(P arrayPtr)
...
+            {
+                if (__ctfe)
+                    return;
+
+                assert(&this !is null);
+                auto s = next;
+                while (s)
+                {
+                    auto sn = s.next;
+                    GC.free(s);
```
Comment (DmitryOlshansky) [inline std/array.d]: And this is GC.free. Something doesn't add up here.

**Comment by JackStouffer** [comment]
Builder is an order of magnitude faster when working with small arrays.

Which is interesting because the argument can that in real world code, arrays are usually smaller than 500 elements. Next is to measure the performance of ranges which put one element at a time.

```
appender	12 ms, 702 μs, and 1 hnsec
builder		1 ms, 81 μs, and 2 hnsecs
```

```d
void main()
{
    foreach (_; 0 .. 200)
        random1 ~= uniform(0, 10_000);
    foreach (_; 0 .. 10)
        random2 ~= uniform(0, 10_000);
    foreach (_; 0 .. 300)
        random3 ~= uniform(0, 10_000);

    auto result  = to!Duration(benchmark!(() => appenderTest())(testCount)[0]);
    auto result2 = to!Duration(benchmark!(() => builderTest())(testCount)[0]);

    writeln("appender", "\t", result);
    writeln("builder", "\t\t", result2);
}
```

File: std/array.d (hunk 2 above)
Comment (JackStouffer) [inline std/array.d]: Heh, that would cause massive GC collections

**Comment by JackStouffer** [comment]
Stupid mistake. Memory leak issue fixed thanks to @DmitryOlshansky 

New performance numbers

Large arrays
```
appender	86 ms, 601 μs, and 1 hnsec
builder		53 ms, 843 μs, and 3 hnsecs
```

Small arrays
```
appender	12 ms, 171 μs, and 4 hnsecs
builder		1 ms, 674 μs, and 4 hnsecs
```

Ranges
```
appender	156 ms, 456 μs, and 5 hnsecs
builder		73 ms, 925 μs, and 4 hnsecs
```

Very promising!

**Comment by DmitryOlshansky** [comment]
> New performance numbers

You should absolutely try `.data` for both Builder and Appender. The builder has to build an array after all not just keep a forward range over chunks of stuff.

**Comment by jondegenhardt** [comment]
If you'd like to run a performance test on an existing app, the tsv-summarize app in the [tsv-utils](https://github.com/eBay/tsv-utils-dlang) project makes significant use of appender for certain calculations, like median. It's a particular use-style, continually appending to the end, then using the data at the end. I can suggest specific tests if you decide you want to try it.

**Comment by JackStouffer** [comment]
New numbers using `builder.data()` rather than returning a slice

```
# large arrays
appender	817 ms, 573 μs, and 9 hnsecs
builder		1 sec, 667 ms, and 160 μs

# small arrays 
appender	118 ms, 54 μs, and 7 hnsecs
builder		66 ms, 241 μs, and 3 hnsecs

# ranges
appender	1 sec, 618 ms, and 797 μs
builder		744 ms, 555 μs, and 7 hnsecs
```

**Comment by JackStouffer** [comment]
LDC makes the builder 15% faster for large arrays while the appender speed stays constant

```
appender	702 ms, 887 μs, and 3 hnsecs
builder		1 sec, 427 ms, 497 μs, and 1 hnsec
```

**Comment by JackStouffer** [comment]
@jondegenhardt Are there any pieces of code in particular I should look at?

**Comment by jondegenhardt** [comment]
@JackStouffer If builder is a drop-in replacement for appender then you could simply replace all the occurrences in the file `tsv-utils-dlang/tsv-summarize/src/tsv-summarize.d`. Build it by cd'ing to the `tsv-utils-dlang/tsv-summarize` directory and running either `make` or `make DCOMPILER=ldc2` (replace ldc2 with a path to your compiler). Run tests using `make test` or `make unittest`. The executable will show up in `tsv-utils-dlang/bin`.

You'll need a reasonably large tsv data file to work with it. You can generate one of course, but a couple that would be good for this test are the google n-gram files, available here: http://storage.googleapis.com/books/ngrams/books/datasetsv2.html. The first 5 files under English 1-grams are about 3 GB, 16 million lines total (googlebooks-eng-all-1gram-20120701-[0-4]).

If you have those try:
```
tsv-summarize --count --median 3,4 googlebooks-eng-all-1gram-20120701-*
```
this will create two appenders of 16 million entries (two values for each line). The command:
```
tsv-summarize --count --group-by 2 --median 3,4 googlebooks-eng-all-1gram-20120701-*
```
will create 850 appenders ranging in size from a few 10s to a bit over 2 million, averaging about 400k. (Field 2 is the 'year' field, there are 425 distinct years.)

These commands take about 40 seconds on my Macbook (ldc 1.1). If appender is materially faster or slower, it'll show up. If you want to focus on smaller array sizes, use fewer files. Use the same phobos version when comparing runs. The median calculation is much faster in 2.073.

**Comment by JackStouffer** [comment]
DMD with appender
```
time ./tsv-summarize --count --median 3,4 /Users/Jack/Downloads/googlebooks-eng-all-1gram-20120701-0 /Users/Jack/Downloads/googlebooks-eng-all-1gram-20120701-1 /Users/Jack/Downloads/googlebooks-eng-all-1gram-20120701-2 /Users/Jack/Downloads/googlebooks-eng-all-1gram-20120701-3 /Users/Jack/Downloads/googlebooks-eng-all-1gram-20120701-4

55.63s user 7.07s system 94% cpu 1:06.41 total
```

DMD with builder
```
time ./tsv-summarize --count --median 3,4 /Users/Jack/Downloads/googlebooks-eng-all-1gram-20120701-0 /Users/Jack/Downloads/googlebooks-eng-all-1gram-20120701-1 /Users/Jack/Downloads/googlebooks-eng-all-1gram-20120701-2 /Users/Jack/Downloads/googlebooks-eng-all-1gram-20120701-3 /Users/Jack/Downloads/googlebooks-eng-all-1gram-20120701-4
160946067	3	3
./tsv-summarize --count --median 3,4       52.07s user 7.60s system 84% cpu 1:10.29 total
```

Will post LDC numbers soon

**Comment by JackStouffer** [comment]
The only conclusion to be reached here is that there is no performance change when builder is used in the manner that tsv-summarize is using appender.

LDC with appender
```
time ./tsv-summarize --count --median 3,4
44.09s user 4.89s system 97% cpu 50.042 total

time ./tsv-summarize --count --group-by 2 --median 3,4
49.57s user 3.80s system 97% cpu 54.639 total
```

LDC with builder
```
time ./tsv-summarize --count --median 3,4
44.92s user 4.66s system 96% cpu 51.260 total

time ./tsv-summarize --count --group-by 2 --median 3,4
53.08s user 5.15s system 97% cpu 59.918 total
```

**Comment by jondegenhardt** [comment]
Thanks for testing this. Perhaps consistent with your earlier numbers for large arrays.

**Comment by JackStouffer** [comment]
Based on further benchmarks and tests, I see no major performance gains using this scheme if https://github.com/dlang/phobos/pull/5162 is pulled. There's really no getting around the `O(n)` `data` performance.

Any potential differences really fall away when the user can make a guess as to the data's length and use `Appender.reserve`.

Also, Appender can be used in safe code while this has way too many unsafe features to rely on trusted lambdas.

**Comment by JackStouffer** [comment]
Closing.

When this was originally proposed, Appender might have been way slower, but that's no longer the case. With LDC's optimizer, I see no difference in performance in real world cases.

Thanks everyone for the feedback.

**Pull request closed without merging**